### PR TITLE
Add support to send notifications to simulator

### PR DESCRIPTION
**Pull request by ojii**
Based off #22, probably shouldn't be merged into that branch but targets it for the cleaner diff. For now just an idea how the api could look like (needs docs & tests etc)

### CODE REVIEW

File: src/aapns/api.py [hunk 1]
```diff
@@ -1,14 +1,111 @@
...
 
 from . import config, errors, models
 from .pool import Pool, Request, create_ssl_context
 
 
-@attr.s(auto_attribs=True, frozen=True)
-class APNS:
-    server: config.Server
+class APNSBaseClient(metaclass=abc.ABCMeta):
+    @abc.abstractmethod
```
Comment (dimaqq) [inline src/aapns/api.py:15]: man you love those abstract classes a tad too much 😓

Comment (dimaqq) [reply src/aapns/api.py:15]: Would you consider PEP-544 "Structural subtyping" instead?
https://www.python.org/dev/peps/pep-0544/

File: src/aapns/api.py [hunk 2]
```diff
@@ -1,14 +1,111 @@
...
+        apns_id: Optional[str] = None,
+        expiration: Optional[int] = None,
+        priority: config.Priority = config.Priority.normal,
+        topic: Optional[str] = None,
+        collapse_id: Optional[str] = None,
+    ) -> Optional[str]:
+        with TemporaryDirectory() as workspace:
+            path = os.path.join(workspace, "notification.apns")
+            with open(path, "wb") as fobj:
+                fobj.write(notification.encode())
```
Comment (dimaqq) [inline src/aapns/api.py:96]: `pathlib` and `.write_text` would make this neater, I think.

**Comment by dimaqq** [comment]
Consider rebasing on top of master :)

File: .circleci/config.yml [hunk 3]
```diff
@@ -11,7 +11,7 @@ commands:
             wget https://raw.githubusercontent.com/sdispater/poetry/master/get-poetry.py
       - run:
           name: Install Poetry
-          command: POETRY_PREVIEW=1 python get-poetry.py  --yes
+          command: python get-poetry.py  --yes
```
Comment (dimaqq) [inline .circleci/config.yml:14]: one space too many in `  --yes` 🦅 👁️

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -1,15 +1,20 @@
 repos:
 - repo: https://github.com/ambv/black
-  rev: stable
+  rev: 19.10b0
   hooks:
   - id: black
     language_version: python3.7
 - repo: https://github.com/timothycrosley/isort
-  rev: 4.3.21-2
+  rev: 4.3.21
```
Comment (dimaqq) [inline .pre-commit-config.yaml:8]: I can't recall what the different was between `.21` and `.21-2`, I think the latter was a fixed github hook, while they both use unsuffixed pypi version...

File: docs/api.rst [hunk 5]
```diff
@@ -1,93 +1,25 @@
 API Reference
 =============
 
-
 ``aapns.api``
 -------------
 
+.. automodule:: aapns.api
```
Comment (dimaqq) [inline docs/api.rst:7]: 😮 😲 🎉

File: docs/api.rst [hunk 6]
```diff
@@ -1,93 +1,25 @@
...
-
-        This is a coroutine.
-
-        Send a notification to the device registered for the given token. Returns
-        the notification ID.
-
-        :param str token: Device token
-        :param notification: The notification to send
-        :type notification: :py:class:`aapns.models.Notification`
-        :param str apns_id: Optional notification ID. If none is provided, APNS creates one
```
Comment (dimaqq) [inline docs/api.rst:36]: How can this extra text be injected into auto module?

File: docs/requirements.txt [hunk 7]
```diff
@@ -0,0 +1 @@
+..
```
Comment (dimaqq) [inline docs/requirements.txt:1]: wow. magic!

File: docs/extended_usage.rst [hunk 8]
```diff
@@ -6,12 +6,12 @@ Quickstart
...
 
-   from aapns.api import create_client
-   from aapns.config import Priority, Production
+   from aapns.api import Server
+   from aapns.config import Priority
    from aapns.models import Notification, Alert, Localized
 
    async def send_hello_world():
-       client = await create_client('/path/to/push/cert.pem', Production)
+       client = await Server.production('/path/to/push/cert.pem').create_client()
```
Comment (dimaqq) [inline docs/extended_usage.rst:14]: Would be nice to document the incompatible change in the change log.

File: docs/api.rst (hunk 6 above)
Comment (ojii) [reply docs/api.rst:36]: I think I  could add comments to attributes to do this, but basically I've opted for "read the apple docs to understand what these do".

File: docs/requirements.txt (hunk 7 above)
Comment (ojii) [reply docs/requirements.txt:1]: needed for autodoc on RTD I think.

**Pull request merged**